### ISSUE TITLE
Are we stalled until next year April?

### ISSUE THREAD

**Issue by zoomclub** [Famous/famous-react#10]
According to the following news from famo.us mixed mode is not really available until next year April. 

http://us5.campaign-archive1.com/?u=4656ba2b0a364690c8530bc1e&id=0c2d8d0d20&e=e99b20f268

This is a long time! Famo.us seems to be good at parading their dreams around and having seeming partners get up on stage with them. This goes beyond hype into the realm of creating FUD in the industry. Another term for this might be crazymaker.

We are seeing a lot of react integrations with everything else now, what is the hold up with famo.us?

**Comment by trusktr** [comment]
@contra Just a suggestion, but when mixed mode is ready, and if you're not already planning this, focus on making react's `createClass` being able to produce a RenderNode. In the return value of `createClass`, let the user return an HTML template like normal, or return a RenderNode.

In the first case, when an HTML element is returned, that cab be easily put inside a sufrace.

In the second case, returning a render node would make it easy to string together components of a RenderTree, since RenderNode is to Famous RenderTree as DOM is to Traditional HTML.

That will allow composability of a RenderTree in the same way that people are composing a DOM with react.

**Comment by trusktr** [comment]
Maybe your `createFamous` method can return an extended node instead of an object containing a node? Just an idea, but that might make things more composable. I was thinking about it. I don't know what you were planning though, maybe you've already got some better ideas. :D

**Comment by trusktr** [comment]
For example, so this is a basic react example:

```js
var CommentBox = React.createClass({
  render: function() {
    return (
      <div className="commentBox">
        Hello, world! I am a CommentBox.
      </div>
    );
  }
});
React.render(
  <CommentBox />,
  document.getElementById('content')
);
```

Following that same basic pattern, here's what it could be like with a RenderNode.

```js
var CommentBox = React.createClass({
  render: function() {
    return 
      new ReactRenderNode({
        id: "foo"
        children: [
          new Surface({ /* data-bound properties in here, for example */ }),
          new ReactRenderNode({ /* ... */ })
        ]
      });
  }
});
React.render(
  CommentBox,
  FamousReact.getNode("bar") // renders the CommentBox tree into the ReactRenderNode that has id "bar"
);
```

So the flow of composing a render tree would be identical to how you compose a DOM in react.

**Comment by trusktr** [comment]
You could change `React.render` to `FamousReact.render `, etc, if you want to, but the idea would be the same.

**Comment by trusktr** [comment]
Here's the composing example.

Regular ol' DOM composing:

```js
var CommentList = React.createClass({
  render: function() {
    return (
      <div className="commentList">
        Hello, world! I am a CommentList.
      </div>
    );
  }
});

var CommentForm = React.createClass({
  render: function() {
    return (
      <div className="commentForm">
        Hello, world! I am a CommentForm.
      </div>
    );
  }
});

var CommentBox = React.createClass({
  render: function() {
    return (
      <div className="commentBox">
        <h1>Comments</h1>
        <CommentList />
        <CommentForm />
      </div>
    );
  }
});
```

With FamousReact, it could be like

```js
var CommentList = React.createClass({
  render: function() {
    return new ReactRenderNode({
      id: "commentList"
      children: [
        new Surface({ /* data-bound properties in here, for example. */
          content: "Hello, world! I am a CommentList."
        })
      ]
    });
  }
});

var CommentForm = React.createClass({
  render: function() {
    return new ReactRenderNode({
      id: "commentForm"
      children: [
        new Surface({ /* data-bound properties in here, for example. */
          content: "Hello, world! I am a CommentForm."
        })
      ]
    });
  }
});

var CommentBox = React.createClass({
  render: function() {
    return new ReactRenderNode({
      id: "commentBox"
      children: [
        new Surface({ /* data-bound properties in here, for example. */
          content: "<h1>Comments</h1>"
        }),
        new CommentList,
        new CommentForm
      ]
    });
  }
});
```

**Comment by trusktr** [comment]
@zoomclub ^ Would an API like that bring you back to Famo.us?

**Comment by trusktr** [comment]
Data binding example:

```js
var HelloMessage = React.createClass({displayName: "HelloMessage",
  render: function() {
    return new Node({
      children: [
        new Surface({ /* data-bound properties in here, for example. */
          content: "Hello " + this.props.name
        })
      ]
    });
  }
});

React.render(FamousReact.createNode(HelloMessage), {name: "John"}), mountNode);
```

**Comment by trusktr** [comment]
Regular DOM components could be put inside surfaces, so when their data changes, only those DOM elements get recreated, not the render nodes.

**Comment by trusktr** [comment]
@zoomclub ^ Would an API like that bring you back to Famo.us?

**Comment by trusktr** [comment]
It would then be possible to take things a step further, with JSX,

```js
var CommentBox = React.createClass({
  render: function() {
    return (
      <Node id="commentBox">
        <Surface>
          <h1>Comments</h1>
        </Surface>
        <CommentList />
        <CommentForm />
      </Node>
    );
  }
});
```

and generate a RenderTree.

**Comment by trusktr** [comment]
@zoomclub ^ Would an API along those lines bring you back to Famo.us?

**Comment by trusktr** [comment]
Or with a Jade plugin:

```js
var CommentBox = React.createClass({
  render: function() {
    return (
      Node(id="commentBox")
        Surface
          h1 Comments on the famous-react API
        CommentList
        CommentForm
    );
  }
});
```

**Comment by trusktr** [comment]
@zoomclub ^ Would an API along those lines bring you back to Famo.us?

**Comment by zoomclub** [comment]
It would bring me and many others back to the fold, this looks fabulous indeed!

I agree that there is no reason to wait for mixed mode to get famo.us together with react. React is happening now and famo.us is too but is driving with the brakes on. What difference does it make that famo.us wants to be the official view layer in the next version of angular? Who cares when angular2 is still far off and react is taking off!

**Comment by stereokai** [comment]
@zoomclub :+1: Totally agree on this. With what projects like React, Mithril and Ractive bring to the table, Angular feels so bloated and outdated... not to mentioned being really hostile to novice devs and beginners.

@trusktr :+1: :+1: :+1: :+1: :+1:  Love your work!

**Comment by coodoo** [comment]
Thanks @trusktr  for the prototype idea, it would be even better if we could make following snippet works directly with existing React syntax (meaning asking React to take gl/famous as a rendering backend, instead of using two different syntax to construct the components)

```js
var CommentBox = React.createClass({
  render: function() {
    return (
      <div className="commentBox">
        Hello, world! I am a CommentBox.
      </div>
    );
  }
});
React.render(
  <CommentBox />,
  document.getElementById('content')
);
```

Do you think it's possible?

**Comment by trusktr** [comment]
Anything is possible, but I think I'd disagree with that. You're better off
learning how the famous RenderTree works and how the DOM works, separately,
so you can have the most control over your app. This would require Famo.us
to completely reimplement the W3C DOM spec, which is not likely to happen,
probably never.

/#!/JoePea

**Comment by coodoo** [comment]
Not even possible to think React/Jsx as a higher level DOM-ish DSL, which already supports a lot of rendering backends like Canvas and SVG, then translate jsx nodes automatically to proper famous equivalent representations? 

Simply put, think it this way: use jsx as the "markup/object graph" language for famous.

The reason behind it is by doing so, it would be much easier to get existing react developers on board and startup using famous, if that aligns with the company's goal.

**Comment by zoomclub** [comment]
Finally, some honesty about angular in the following link. Hope famo.us takes note and stops looking to supports like angular and joins up with winners like react soon. Enough of the big frameworks and pie in the sky dreams of retooling everything, lets just have some great libraries that work well together already.

http://www.quirksmode.org/blog/archives/2015/01/the_problem_wit.html

**Comment by trusktr** [comment]
@coodoo Did you miss my idea above? I'll write it here again for reference:

```js
var CommentBox = React.createClass({
  render: function() {
    return (
      <Node id="commentBox">
        <Surface>
          <h1>Comments</h1>
        </Surface>
        <CommentList />
        <CommentForm />
      </Node>
    );
  }
});
```

That would be the `"markup/object graph"` you're talking about. ;)

**Comment by stereokai** [comment]
@zoomclub Finally? Haven't you heard about this wave-making article? http://www.fse.guru/2-years-with-angular

Anyway, thanks for your link, you sent me on a 2-day reading frenzy, on lots of web development, frameworks and programming topics. One of the things I'd like to take that relate to your comment - I believe AngularJS is the next jQuery. The future lies in virtual dom.

**Comment by petehunt** [comment]
relevant: https://www.youtube.com/watch?v=eNC0mRYGWgc

**Comment by coodoo** [comment]
Now that React Native is out in the wild which not only proved the idea I mentioned is totally doable (retargeting jsx toward different drawing backend), it also rendered this issue no longer relevant :)

Thanks Pete and co@fb for the good job!

**Comment by marcus** [comment]
@trusktr Your examples of React / Famo.us / JSX integration are *exactly* what I was out looking for. Thanks for taking the time to prototype those. If they became a reality, using Famo.us would almost be an easy decision for me and my team.

The most frustrating thing about famo.us to me is that it is almost impossible to figure out the state of the project. Mixed mode sounds cool, but details are so sparse, and so hard to dig up, that despite the reassuring message that the team isn't going to pull an Angular 2.0, I'm still hesitant because I can't tell where the message is coming from.

I'm not asking for code to be released or a set in stone timeline or anything, just clearer, easier to find messaging. The famo.us blog, for example, is full of shaky videos showing what appears to be awesome progress, but at the same time all these demos are being produced, the demo link disappears from the front page of the website and I'm back to googling for old projects to try to get buy-in from coworkers. 

Another example - For what it's worth, to me, the jQuery announcement was just strange. The integration seems to be nothing but a publicity stunt since the announcement has no technical details. I can't see why you'd want/need jQuery integration and the press release has no details on that. Everything famo.us does seems to be hidden in 3rd party articles from places like Forbes or scattered across YouTube videos. An up to date roadmap would be fantastic. A list of current demos. Information on the progress of mixed-mode. Anything.

I'm not even sure where to post this message where it will be read. The only community around famo.us seems to be the IRC (irc?!) channel linked to from the help page. Basically, famo.us seems amazing, but I just can't tell where it's going. It's frustrating. Sorry for hijacking this thread with a random rant and again, thanks @trusktr .

**Comment by trusktr** [comment]
@marcus Checkout http://blitzagency.com. They made a Backbone/Marionette/Famo.us integration called [rich](https://github.com/blitzagency/rich), which is what the site is made with.

Checkout famous-views.meteor.com, a super nice integration with Meteor.

[Superstereo](http://superstereo.com) isn't released yet, but it's made with Famo.us, and it's looking amazing.

At the end of Demo Day 10, at [25:40](https://www.youtube.com/watch?v=lxvuHtYGgcM#t=1540), the power of Mixed Mode becomes extremely apparent. There's no words to describe how amazing that feat is.

In it's current state, Famo.us is powerful. Just give it time and more amazing things will start showing up in the wild. I, for example, have adopted Famo.us since the moment it came out, but I have not yet released my personal project to production because it just *takes time*. Don't wait for Famo.us to release integrations or more documentation, dive into the code, learn it's amazing power, and start creating amazing things. It's completely doable right now, without needing Mixed Mode. Things will just be even better when Mixed Mode is released.

Also! Famo.us just released the new [Reference Documentation](http://famo.us/docs/reference/pages/0.3/index.html). It's a really nice addition to the docs, and it even includes Physics examples!

**Comment by trusktr** [comment]
@contra If you'll be working on this in the near future, please don't be blocked by the wait for Mixed Mode. My ideas above can be implemented using the current state of Famo.us, and won't require many changes for Mixed Mode. The render tree isn't going anywhere.

**Comment by contra** [comment]
@trusktr We were told by famous to stop working until mixed mode comes out, so that's what we are going to do. I have no estimates on when it will be ready so we are working on other things until then.
